google-cloud-spanner/src/main/java/com/google/cloud/spanner/TransactionRunnerImpl.java

```diff
@@ -479,6 +491,32 @@ public void run() {
                     return;
                   }
                   com.google.spanner.v1.CommitResponse proto = commitFuture.get();
+
+                  // If the CommitResponse includes a precommit token, the client will retry the
+                  // commit RPC once with the new token and clear any existing mutations.
+                  // This case is applicable only when the read-write transaction uses multiplexed
+                  // session.
+                  if (proto.hasPrecommitToken() && !retryAttemptDueToCommitProtocolExtension) {
+                    // track the latest pre commit token
+                    onPrecommitToken(proto.getPrecommitToken());
+                    span.addAnnotation(
+                        "Commit operation will be retried with new precommit token as the CommitResponse includes a MultiplexedSessionRetry field");
+                    opSpan.addAnnotation(
+                        "Commit operation will be retried with new precommit token as the CommitResponse includes a MultiplexedSessionRetry field");
+                    opSpan.end();
+
+                    // Retry the commit RPC with the latest precommit token from CommitResponse.
+                    MoreExecutors.directExecutor()
+                        .execute(
+                            new CommitRunnable(
+                                res,
+                                prev,
+                                requestBuilder,
+                                /* retryAttemptDueToCommitProtocolExtension = */ true));
```

Comment:
We should preferably not call this here:
1. This is just a complicated way of executing `new CommitRunnable(...).run()`.
2. That is a blocking call that will execute a new RPC.
3. We should not do that in the listener of the previous commit, as this will then be executed by a gRPC worker thread. That thread will then be blocked while waiting for this second commit to finish.

I'm not entirely sure what the best way is to bring this retry out of this listener, but I think that you need to return a specific error here, and then have the caller of this method handle that error. That handling would then be 'retry the commit with the new pre-commit token'.

Reply:
Per offline discussion: `CommitRunnable.run()` seems to be non-blocking. That would mean that this is safe to do.